### PR TITLE
fix(speedup): refactor the execution logic of functions in speedup

### PR DESCRIPTION
**Pull request by Louis-J**
### Description ###

In speedup, we use a non-general trick to treat the prim ops. We analyze the relationship between the prim ops, then delete all the prim ops, finally connect the non-prim op directly. We also assume that all the intermediate variables are tensors, tuples of tensors, or lists of tensors. But it'll cause many problems in various models.

So, it's better to execute all the prim ops for real. In this pr, I remove the deletion of prim ops, and execute the prim recursively. And the assumption of variable types is also removed. Now it's closer to the original execution.

But it changed a lot and is untested now. More work is needed.

#5097

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: docs/source/compression/pruning.rst [hunk 1]
```diff
@@ -105,4 +105,5 @@ In addition, for the convolutional layers that have more than one filter group,
 ``dependency-aware pruner`` will also try to prune the same number of the channels for each filter group.
 Overall, this pruner will prune the model according to the L1 norm of each filter and try to meet the topological constrains (channel dependency, etc) to improve the final speed gain after the speedup process. 
 
+Operations that will be recognized as having channel dependencies: add/sub/mul/div, addcmul/addcdiv, logical_and/or/xor
```
Comment (J-shang) [inline docs/source/compression/pruning.rst:108]: `.`?

File: nni/common/graph_utils.py [hunk 2]
```diff
@@ -725,8 +717,6 @@ def _build_graph(self):
 
         # associate module name with their trace graph nodes
         for node in graph.nodes():
-            if node.kind() == CONSTANT_KIND:
-                continue
```
Comment (J-shang) [inline nni/common/graph_utils.py:729]: when we should skip the CONSTANT_KIND and when we should not? I found you did not remove all the `if node.kind() == CONSTANT_KIND:` in the code

File: nni/compression/pytorch/speedup/compressor.py [hunk 3]
```diff
@@ -175,12 +176,15 @@ def _prepare_dummy_input(self, node):
         # prepare the inputs and outputs mask for this node,
         # if there is already a mask in self.masks, then use
         # the original mask tensor, else create a new one.
-        inputs_name = node.inputs
+        if node.type == 'module':
+            inputs_name = node.inputs
+        else:
+            inputs_name = [val_node.debugName() for val_node in node.key_node.inputs()]
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/compressor.py:182]: what is key_node? maybe we need a doc to explain the important attrs in a node

File: nni/compression/pytorch/speedup/compressor.py [hunk 4]
```diff
@@ -297,16 +318,15 @@ def update_indirect_sparsity(self, node):
                 debug_name = auto_infer.input_debugname[in_id]
 
                 last_output = self.internal_result[debug_name]
-                # if isinstance(last_output, torch.Tensor):
-                # TODO what if last output is tuple/list of tensor
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/compressor.py:301]: so `what if last output is tuple/list of tensor`?

**Comment by Louis-J** [comment]
totally refactored old speedup in https://github.com/Louis-J/nni/tree/refactor_speedup

**Comment by Louis-J** [comment]
the speedup v2 now is runnable. so better to reassess that refactor the old speedup is suitable or not.

**Comment by Lijiaoa** [comment]
https://github.com/microsoft/nni/issues/5143

**Pull request closed without merging**